Adding slack sink connector

Kanugula opened the pull request:
PR closed by Hazelcast automation as no activity (>6 months). Please reopen with comments, if necessary. Thank you for using Hazelcast and your valuable contributions

Kanugula committed:
Adding slack sink connector

devOpsHazelcast commented:
Can one of the admins verify this patch?

cangencer commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,90 @@
+     * @param channel String Unique channel id to send messages to the slack channel.
+     * @return
+     */
+    public static Sink sink(String accessToken, String channel) {
```
can we call this `slackChannel`  instead?

Kanugula replied:
@cangencer Sure. I will update the variable name to slackChannel in the next patchset.

cangencer commented:
Thanks a lot for the contribution! I couldn't see your name in our contributor list: could you follow the steps at https://hazelcast.atlassian.net/wiki/spaces/COM/pages/6357071/Hazelcast+Contributor+Agreement ?

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,90 @@
+     * @return
+     */
+    public static Sink sink(String accessToken, String channel) {
+        return SinkBuilder.sinkBuilder("slack-sink",
```
Would be useful to include the channel name in the sink name.

Kanugula replied:
The new method has been added to the file with the channel name as 3rd parameter.

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,90 @@
+            request.setEntity(new UrlEncodedFormEntity(urlParameters));
+            CloseableHttpResponse response = (CloseableHttpResponse) closeableHttpClient.execute(request);
+            String result = EntityUtils.toString(response.getEntity());
+            request.releaseConnection();
```
The release should be in a `finally` block

Kanugula replied:
Yes Request and other objects are closed in `finally` block

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,90 @@
+            urlParameters.add(new BasicNameValuePair("channel",  channel));
+            urlParameters.add(new BasicNameValuePair("text", message));
+            request.setEntity(new UrlEncodedFormEntity(urlParameters));
+            CloseableHttpResponse response = (CloseableHttpResponse) closeableHttpClient.execute(request);
```
A production grade connector should tolerate failures. The only way to do so in `SinkBuilder` is to retry in a loop and don't return until the message can be sent. We should also possibly reconnect on a failure.

I think we'll merge the sink in current state, but it should be stated in its javadoc that on any connection failure it will cause the job to fail.

Kanugula replied:
Updated the code with HttpRetryHandler. Some of the exceptions are handled. Remaining exceptions are rethrown using ExceptionUtils.

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    /**
+     * Creates a sink to send messages to the slack channel>
```
Typo: `>` at the end.

Kanugula replied:
Updated.

eminn commented:
verify

Kanugula committed:
* Added additional method with slackChannelName to name the sink.
* Added retry http handler to handle some of the error causes.
* Variable names are updated as per review comments.
* Addressed other review comments.

Signed-off-by: Rajasekhar Kanugula <Rajasekhar.Kanugula@gmail.com>

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+     * @return
+     */
+    public static Sink sink(String accessToken, String slackChannelId, String slackChannelName) {
+        return SinkBuilder.sinkBuilder(slackChannelName,
```
It's not `slackChannelName`, but `sinkName`. But my point was that we don't need the user to ask about the sink name. We could generate it like this: `"slack(" + slackChannelId + ")"`

Kanugula replied:
@viliam-durina Remoed the slackChannelName attribute and named the sink as you said.

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+                throw ExceptionUtil.rethrow(var);
+            } catch (ClientProtocolException var) {
+                throw ExceptionUtil.rethrow(var);
+            } catch (IOException var) {
```
You can combine these three:
```java
} catch (UnsupportedEncodingException | ClientProtocolException | IOException var) {
```

Kanugula replied:
@viliam-durina `UnsupportedEncodingException  & ClientProtocolException ` are chile classes of `IOException `. Updated the changes here.

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+            }
+            if (exception instanceof UnknownHostException) {
+                // Unknown host
+                return false;
```
Host can be unknown temporarily, just like connection timeout.

Kanugula replied:
Handled the UnknownHostException also.

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+        return (exception, executionCount, context) -> {
+            if (executionCount >= RETRY_COUNT) {
+                // Do not retry if over max retry count
+                return false;
```
I think we can retry forever. If you have the slack app, it also retries forever.

Kanugula replied:
I thought retrying forever might cause issues related to stability. Now I updated the retry count to Integer.max value instead of 3 times.

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+                urlParameters.add(new BasicNameValuePair("text", message));
+                request.setEntity(new UrlEncodedFormEntity(urlParameters));
+                response = (CloseableHttpResponse) closeableHttpClient.execute(request);
+                result = EntityUtils.toString(response.getEntity());
```
We return the result, but the returned value is ignored. Is there something worthwhile to check?

Kanugula replied:
Added the simple to pojo class to parse the response. In error scenarios sample message will look like this `{
    "ok": false,
    "error": "not_authed"
}` If the Response.ok false means then throwing the RuntimeException.

viliam-durina commented on slack/src/test/java/com/hazelcast/jet/contrib/slack/SlackSInksTest.java:
```diff
@@ -0,0 +1,155 @@
+import java.util.Map;
+import java.util.Properties;
+
+public final class SlackSInksTest extends JetTestSupport {
```
"SInks" - upper case I

Kanugula replied:
Updated the class name

viliam-durina commented on slack/src/test/java/com/hazelcast/jet/contrib/slack/SlackSInksTest.java:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    @Test
+    public void retryTests() {
```
This is probably your development test code, not a real unit test.

Kanugula replied:
Yes development test code and it's removed from the test class

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+     * custom {@link HttpRequestRetryHandler} to post the messages to slack channels.
+     * Following error causes {@link ConnectTimeoutException}, {@link HttpHostConnectException}, {@link IOException}
+     * are handled through custom retry handler.The max retry count has been configured to 5.
+     * Other error causes will lead to the breakage of the jet pipeline.
```
Multiple issues:
- wrap the lines with 72 characters
- grammar: use capital letters for proper names (Apache, Slack), use space after a period.
- "Following error causes" -> "Following exceptions"
- "breakage of the jet pipeline" -> "failure of the Jet job"

Kanugula replied:
Updated

viliam-durina commented on slack/src/main/java/com/hazelcast/jet/contrib/slack/SlackSinks.java:
```diff
@@ -0,0 +1,170 @@
+    /**
+     * Creates a slack sink to send messages to the slack channel with the given slack channel name.
+     * Internally apache {@link org.apache.http.client.HttpClient} is used with
+     * custom {@link HttpRequestRetryHandler} to post the messages to slack channels.
```
"to slack channels" - do we support sending to multiple channels? Otherwise it should be "to a Slack channel".

Kanugula replied:
right now we are sending messages to a single channel only. Updated as mentioned

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org:443/pull/badge/signed)](https://cla.hazelcast.org:443/hazelcast/hazelcast-jet-contrib?pullRequest=60) <br/>All committers have signed the CLA.

Holmistr commented:
Hi @Kanugula , just checking in if there's anything we can help you out with? :)

Kanugula committed:
> Updated the documentation as per the review comments.
> Processing the Rest response to SlackResponse to handle failure causes from the slack apis.

Signed-off-by: Rajasekhar Kanugula <Rajasekhar.Kanugula@gmail.com>

Holmistr commented:
@Kanugula Glad to see your edits :) @viliam-durina Could you please take another look?

cangencer commented:
@viliam-durina do you mind picking this up again?
